File: lib/domain/shared/date_calendar.rb
# frozen_string_literal: true

module Domain
  module Shared
    # ActiveSupport の Date 拡張（beginning_of_month / end_of_year 等）に依存しないカレンダー補助。
    module DateCalendar
      module_function

      def beginning_of_month(date)
        Date.new(date.year, date.month, 1)
      end

      # @return [Date] 当月最終日（Ruby 標準: 日にちに -1 を指定）
      def end_of_month(date)
        Date.new(date.year, date.month, -1)
      end

      def beginning_of_year(date)
        Date.new(date.year, 1, 1)
      end

      def end_of_year(date)
        Date.new(date.year, 12, 31)
      end

      # 翌月1日（当月任意の日から）
      def first_day_of_next_calendar_month(date)
        end_of_month(date) + 1
      end
    end
  end
end

File: crates/agrr-domain/src/shared/helpers/date_calendar.rs
//! Ruby: `Domain::Shared::DateCalendar` — calendar helpers without ActiveSupport.

use time::{Date, Month};

/// Ruby: `DateCalendar.beginning_of_month`
pub fn beginning_of_month(date: Date) -> Date {
    Date::from_calendar_date(date.year(), date.month(), 1).expect("valid month start")
}

/// Ruby: `DateCalendar.end_of_month`
pub fn end_of_month(date: Date) -> Date {
    first_day_of_next_calendar_month(date) - time::Duration::days(1)
}

/// Ruby: `DateCalendar.beginning_of_year`
pub fn beginning_of_year(date: Date) -> Date {
    Date::from_calendar_date(date.year(), Month::January, 1).expect("valid year start")
}

/// Ruby: `DateCalendar.end_of_year`
pub fn end_of_year(date: Date) -> Date {
    Date::from_calendar_date(date.year(), Month::December, 31).expect("valid year end")
}

/// Ruby: `DateCalendar.first_day_of_next_calendar_month`
pub fn first_day_of_next_calendar_month(date: Date) -> Date {
    let y = date.year();
    let (next_y, next_m) = if date.month() == Month::December {
        (y + 1, Month::January)
    } else {
        let next_m = Month::try_from(u8::from(date.month()) + 1).expect("month");
        (y, next_m)
    };
    Date::from_calendar_date(next_y, next_m, 1).expect("valid next month start")
}

#[cfg(test)]
mod helpers_date_calendar_test_inline {
    use super::*;
    include!(concat!(env!("CARGO_MANIFEST_DIR"), "/test/shared/helpers_date_calendar_test.rs"));
}
